Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -20,13 +21,31 @@ General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finit
 * `ValuedCSP.Instance`: An instance of a VCSP problem over given template.
 * `ValuedCSP.Instance.evalSolution`: An evaluation of the VCSP instance for given solution.
 * `ValuedCSP.Instance.IsOptimumSolution`: Is given solution a minimum of the VCSP instance?
+* `Function.HasMaxCutProperty`: Can given binary function express the Max-Cut problem?
+* `FractionalOperation`: Multiset of operations on given domain of the same arity.
+* `FractionalOperation.IsSymmetricFractionalPolymorphismFor`: Is given fractional operation a
+   symmetric fractional polymorphism for given VCSP template?
 
 ## References
 * [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
    *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
 
 -/
 
+lemma column_of_2x2_left {α : Type*} (a b c d : α) :
+    (fun i => ![![a, b], ![c, d]] i 0) = (fun i => ![a, c] i) := by
+  ext i
+  fin_cases i <;> rfl
+
+lemma column_of_2x2_right {α : Type*} (a b c d : α) :
+    (fun i => ![![a, b], ![c, d]] i 1) = (fun i => ![b, d] i) := by
+  ext i
+  fin_cases i <;> rfl
+
+lemma univ_sum_2x2_aux {α β : Type*} {f : (Fin 2 → α) → β} [AddCommMonoid β] {a b c d : α} :
```

Comment:
Seeing as how this already exists, this should be deleted

Reply:
Unfortunately, it cannot be plugged into my proof directly. I need a different (yet defeq) variant.

Reply:
You could use 
`  simp only [Fin.sum_univ_two, Matrix.cons_val_zero, Matrix.cons_val_one, Matrix.head_cons, ← mcfab.left, ← two_nsmul] at contr`

Reply:
Thanks for the suggestion! I nevertheless want to have it as a lemma. This should be an atomic step for the "big" proof.

Reply:
How about using 
```
  rw [Fin.sum_univ_two] at contr
  dsimp at contr
  rw [← mcfab.left, ← two_nsmul] at contr
```
in the proof below?

Alternatively, you can generalize this to `Finset.univ.sum (fun i => f (![x, y] i)) = f x + f y` and move directly under `Fin.sum_univ_two`.

Reply:
I didn't see the last two comments yet. Feel free to pursue my "Alternatively".

Reply:
I don't want to make the most complicated proof more complicated than it already is. I will keep it as a separate lemma. Feel free to suggest a different name for the lemma.

Reply:
> Alternatively, you can generalize this to Finset.univ.sum (fun i => f (![x, y] i)) = f x + f y and move directly under Fin.sum_univ_two.

I would have to provide a version for products, right?

Reply:
All right, I added the generalized version as you suggested.